src/IO/ReadHelpers.h

```diff
@@ -909,15 +909,28 @@ inline T parseFromString(std::string_view str)
 }
 
 
-template <typename ReturnType = void>
+template <typename ReturnType = void, bool dt64_mode = false>
 ReturnType readDateTimeTextFallback(time_t & datetime, ReadBuffer & buf, const DateLUTImpl & date_lut);
 
 /** In YYYY-MM-DD hh:mm:ss or YYYY-MM-DD format, according to specified time zone.
   * As an exception, also supported parsing of unix timestamp in form of decimal number.
   */
-template <typename ReturnType = void>
+template <typename ReturnType = void, bool dt64_mode = false>
 inline ReturnType readDateTimeTextImpl(time_t & datetime, ReadBuffer & buf, const DateLUTImpl & date_lut)
 {
+    static constexpr bool throw_exception = std::is_same_v<ReturnType, void>;
+
+    if constexpr (!dt64_mode)
+    {
+        if (!isNumericASCII(*buf.position()))
+        {
+            if constexpr (throw_exception)
+                throw ParsingException(ErrorCodes::CANNOT_PARSE_DATETIME, "Cannot parse datetime");
+            else
+                return false;
+        }
+    }
+
     /// Optimistic path, when whole value is in buffer.
     const char * s = buf.position();
```

Comment:
Hm, why this place does does not check that `s` has enough data? And I don't see any check upper on the stack...
And this is actually the reason why `01179_insert_values_semicolon` "works", since it assumes that it has 19 bytes (`date_time_broken_down_length`), while it definitely not.

Reply:
```sql
local :) INSERT INTO test_01179 values ('2020-01-01 0'); ('0')
Ok.

local :) INSERT INTO test_01179 values ('2020-01-01 0'); (' ')
Ok.

local :) INSERT INTO test_01179 values ('2020-01-01 0'); ('0 )
Ok.
Exception on client:
Code: 33. DB::Exception: Cannot read data after semicolon: While executing ValuesBlockInputFormat: data for INSERT was parsed from query. (CANNOT_READ_ALL_DATA), Stack trace (when copying this message, always include the lines below):

0. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/exception:134: Poco::Exception::Exception(String const&, int) @ 0x000000001521f5d2 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
1. /src/ch/clickhouse/src/Common/Exception.cpp:103: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000b8309f7 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
2. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/string:1499: DB::Exception::Exception<char const (&) [33]>(int, char const (&) [33]) @ 0x0000000006d55cb8 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
3. /src/ch/clickhouse/src/Processors/Formats/Impl/ValuesBlockInputFormat.cpp:0: DB::ValuesBlockInputFormat::readSuffix() @ 0x00000000124d214d in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
4. /src/ch/clickhouse/src/Processors/Formats/Impl/ValuesBlockInputFormat.cpp:0: DB::ValuesBlockInputFormat::generate() @ 0x00000000124d14ff in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
5. /src/ch/clickhouse/src/Processors/Chunk.h:90: DB::ISource::tryGenerate() @ 0x00000000122e03d8 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
6. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/optional:344: DB::ISource::work() @ 0x00000000122e0089 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
7. /src/ch/clickhouse/src/Processors/Executors/ExecutionThreadContext.cpp:0: DB::ExecutionThreadContext::executeTask() @ 0x00000000122f5d8d in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
8. /src/ch/clickhouse/src/Processors/Executors/PipelineExecutor.cpp:273: DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x00000000122eda90 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
9. /src/ch/clickhouse/src/Processors/Executors/PipelineExecutor.cpp:239: DB::PipelineExecutor::executeImpl(unsigned long, bool) @ 0x00000000122ed11c in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
10. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:274: DB::PipelineExecutor::execute(unsigned long, bool) @ 0x00000000122ecf02 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
11. /src/ch/clickhouse/src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:0: void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x00000000122f8c78 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
12. /src/ch/clickhouse/base/base/../base/wide_integer_impl.h:809: ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000000b901f42 in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
13. /src/ch/clickhouse/contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000b905dae in /src/ch/clickhouse/.cmake-llvm16/programs/clickhouse
14. ? @ 0x00007ffff7e3a9eb in ?
15. ? @ 0x00007ffff7ebe7cc in ?
```

Reply:
But to make the last query throw you need to patch `PeekableReadBuffer::hasUnreadData()` to check that there is no data left